ext/opcache/jit/zend_jit.c

```diff
@@ -2656,6 +2696,198 @@ static zend_lifetime_interval** zend_jit_allocate_registers(const zend_op_array
 	return NULL;
 }
 
+#else /* ZEND_JIT_IR */
+
+static void zend_jit_allocate_registers(zend_jit_ctx *ctx, const zend_op_array *op_array, zend_ssa *ssa)
+{
+	void *checkpoint;
+	int candidates_count, i;
+	zend_jit_reg_var *ra;
+
+	checkpoint = zend_arena_checkpoint(CG(arena));
+	ra = zend_arena_calloc(&CG(arena), ssa->vars_count, sizeof(zend_jit_reg_var));
+	candidates_count = 0;
+	for (i = 0; i < ssa->vars_count; i++) {
+		if (zend_jit_may_be_in_reg(op_array, ssa, i)) {
+			ra[i].ref = IR_NULL;
+			candidates_count++;
+		}
+	}
+	if (!candidates_count) {
+		zend_arena_release(&CG(arena), checkpoint);
+		return;
+	}
+
+	if (JIT_G(opt_flags) & ZEND_JIT_REG_ALLOC_GLOBAL) {
+		/* Naive SSA resolution */
+		for (i = 0; i < ssa->vars_count; i++) {
+			if (ssa->vars[i].definition_phi && !ssa->vars[i].no_val) {
+				zend_ssa_phi *phi = ssa->vars[i].definition_phi;
+				int k, src;
+
+				if (phi->pi >= 0) {
+					src = phi->sources[0];
```

Comment:
I'm still wrapping my head around this, but is it possible for `src` to be a non-existent index on `ra`?

Reply:
No.
``ra[]`` is an array of size ``ssa->vars_count``.
``src`` as a SSA variable index, that must be in range ``[0..ssa->vars_count-1]``.